File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {// tc is o(N) and sc is O(N) using an hashset but we can optimize the space using an bst iterator we can get an Tc is O(N) and sc is O(1)
    public boolean findTarget(TreeNode root, int k) {
        Set<Integer> set = new HashSet<>();
        return dfs(root,k,set);
    }
    private boolean dfs(TreeNode node,int k,Set<Integer> set){
        if(node == null) return false;

        if(set.contains(k-node.val)) return true;
        set.add(node.val);
        return dfs(node.left,k,set) || dfs(node.right,k,set);


    }
}

File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {boolean}
 */
var findTarget = function(root, k) { // tc is O(N) and sc is O(N)
    const set = new Set();
    return dfs(root,k,set);
};
function dfs(node,k,set){
    if(!node) return false;
    if(set.has(k-node.val)) return true;
    set.add(node.val);
    return dfs(node.left,k,set) || dfs(node.right,k,set);
}